### ISSUE TITLE
Join node incorrectly finishes when creating a merged object

### ISSUE THREAD

**Issue by joshchamberlain** [node-red/node-red#1782]
When using a 'join' node to manually 'combine each complete object' to create a 'merged object', the output is sent unreliably. At the end of this ticket, I explain what I think is causing this.

### What are the steps to reproduce?
 1. Copy the flow at the end of this ticket. (Pasted at end for skimability)
 2. Press the top inject node - Injects `a:b:c`
 3. Press the bottom inject node - Injects `b:c:e:i:z`
 4. Press the top inject node - Injects `a:b:c` again

This should send a message to the debug node with the following flags set depending on the injected string:
  * `has_vowel: "true"`: If the string has a vowel (true for top and bottom injector)
  * `has_e_or_i: "true"`: If the string has an e or i (true for the bottom injector)

### What happens?
After step 2:
  * 3 output messages are printed, instead of just 1

After step 3:
  * 2 output messages are printed, instead of just 1

After step 4:
  * 0 output messages are printed, instead of 1

### What do you expect to happen?
After Step 2, the output should be a single message with `has_vowel` set to `true`.

After Step 3, the output should be a single message with `has_vowel` set to `true`, as well as `has_e_or_i` set to `true`.

After Step 4, the output should be identical to that from Step 2.

### Please tell us about your environment:

- [x] Node-RED version: v0.18.7-git
- [x] node.js version: v8.9.4
- [x] npm version: 5.6.0
- [x] Platform/OS: Linux/Ubuntu 16.04
- [x] Browser: Firefix 60.0.2

```
[{"id":"ebf50582.2f74c","type":"inject","z":"e13b2c1d.899d2","name":"Test String With Vowel","topic":"","payload":"a:b:c","payloadType":"str","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":220,"y":140,"wires":[["318775da.d29c8a"]]},{"id":"9585a45f.9a9b28","type":"inject","z":"e13b2c1d.899d2","name":"Test String Without Vowel","topic":"","payload":"b:c:d:f","payloadType":"str","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":230,"y":180,"wires":[["318775da.d29c8a"]]},{"id":"318775da.d29c8a","type":"split","z":"e13b2c1d.899d2","name":"","splt":":","spltType":"str","arraySplt":1,"arraySpltType":"len","stream":false,"addname":"","x":500,"y":160,"wires":[["40ffe89b.d066b"]]},{"id":"40ffe89b.d066b","type":"switch","z":"e13b2c1d.899d2","name":"Check Vowel","property":"payload","propertyType":"msg","rules":[{"t":"eq","v":"a","vt":"str"},{"t":"eq","v":"e","vt":"str"},{"t":"eq","v":"i","vt":"str"},{"t":"eq","v":"o","vt":"str"},{"t":"eq","v":"u","vt":"str"},{"t":"else"}],"checkall":"true","repair":false,"outputs":6,"x":670,"y":160,"wires":[["ea15f264.3b63f8"],["c988f143.71fdf"],["c988f143.71fdf"],["ea15f264.3b63f8"],["ea15f264.3b63f8"],["37420488.5aa77c"]]},{"id":"ea15f264.3b63f8","type":"change","z":"e13b2c1d.899d2","name":"Set has_vowel","rules":[{"t":"set","p":"has_vowel","pt":"msg","to":"true","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":980,"y":140,"wires":[["37420488.5aa77c"]]},{"id":"37420488.5aa77c","type":"join","z":"e13b2c1d.899d2","name":"","mode":"custom","build":"merged","property":"","propertyType":"full","key":"topic","joiner":"\\n","joinerType":"str","accumulate":false,"timeout":"","count":"","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"","reduceFixup":"","x":1150,"y":180,"wires":[["c7f9c2cb.d4f228"]]},{"id":"c7f9c2cb.d4f228","type":"debug","z":"e13b2c1d.899d2","name":"Result","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"true","x":1290,"y":160,"wires":[]},{"id":"c988f143.71fdf","type":"change","z":"e13b2c1d.899d2","name":"Has E or I","rules":[{"t":"set","p":"has_e_or_i","pt":"msg","to":"true","tot":"str"}],"action":"","property":"","from":"","to":"","reg":false,"x":840,"y":80,"wires":[["ea15f264.3b63f8"]]},{"id":"56bfe531.8027c4","type":"inject","z":"e13b2c1d.899d2","name":"Test String With ei","topic":"","payload":"b:c:e:i:z","payloadType":"str","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":210,"y":220,"wires":[["318775da.d29c8a"]]}]
```

### Potential Cause
I think the issue comes from these lines: (`nodes/core/logic/17-split.js` lines 616-617)
```javascript
    group.currentCount = Object.keys(group.payload).length;
    //group.currentCount++;
```
This is in `JoinNode:on("input")` and is used to determine when enough segments have been joined. As far as I can tell, the commented out version is correct. The following fixes the issue:
```javascript
    group.currentCount++;
```
However, it looks extremely intentional, so I'm hesitant to create a pull request.

**Comment by dceejay** [comment]
Yeah - it breaks in other ways when there... when joining a full msg object (fails test) - so needs proper investigation

**Comment by dceejay** [comment]
think it may be down the mix of split and join.... it's splitting a string into 3 or 5 parts - then the join is now trying to create an object - but is using that previous knowledge to say how many parts it needs... so it says 3 properties of msg is ok... etc... so basically every message has more than taht number of properties so it fires on every message...

Need to think what is the sensible way to handle - or if indeed it can be.

**Comment by dceejay** [comment]
I think we probably need to ignore any incoming parts count if the join is of a different type than the split. (as is the case here) - but allow it if it's the same or similar type.

However - as well as that... I think the other deficiency is that maybe the node ought to have worked in automatic mode - and just rejoined the strings back together - that way the parts would have worked and all would be good.... but for the fact the extra properties you added to msg get dropped...  so in fact that may be the real problem.

**Comment by dceejay** [comment]
Have added a post to the forum to discuss fix 
https://discourse.nodered.org/t/proposed-change-to-join-node-default-behaviour/1265

**Comment by joshchamberlain** [comment]
I will be offline for the next week or so, and I have only had time to skim this. Thank you for looking into this, and I'll try to see if I have time to contribute late next week.
